File: leetcode/c++/32. Longest Valid Parentheses.cpp
#include <iostream>
#include <string>
#include <stack>
#include <cmath>

using namespace std;

int longestValidParentheses(string s){
    int maxans=0;
    stack<int> stack;
    for(int i=0;i<s.length();i++){
      if(s[i]=='('){
        stack.push(i);
      }else{
        stack.pop();
        if(stack.empty()){
          stack.push(i);
        }else{
          maxans = max(maxans,i-stack.top());
        }
      }
    }
    return maxans;
}

int main(){
  cout << longestValidParentheses("(()()")<<endl;
  return 0;
}

File: leetcode/python/32. Longest Valid Parentheses.py
def longestValidParentheses(s):
  maxans = 0
  stack = []
  stack.append(-1)
  for i in range(len(s)):
    if s[i] == '(':
      stack.append(i)
    else:
      stack.pop()
      if len(stack)==0:
        stack.append(i)
      else:
        maxans = max(maxans,i-stack[-1])
  return maxans

print(longestValidParentheses("(()()"))